src/rust/engine/src/console_ui.rs

```diff
@@ -0,0 +1,192 @@
+    let multi_progress_bars = MultiProgress::new();
+
+    let bars = (0..num_swimlanes)
+      .map(|_n| {
+        let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+        multi_progress_bars.add(ProgressBar::new(50).with_style(style))
+      })
+      .collect();
+
+    (multi_progress_bars, bars)
+  }
+
+  fn get_label_from_heavy_hitters<'a>(
+    tasks_to_display: impl Iterator<Item = (&'a String, &'a Option<Duration>)>,
+  ) -> Vec<String> {
+    tasks_to_display
+      .map(|(label, maybe_duration)| {
+        let duration_label = match maybe_duration {
+          None => "(Waiting) ".to_string(),
+          Some(duration) => {
+            let duration_secs: f64 = (duration.as_millis() as f64) / 1000.0;
+            format!("{:.2}s ", duration_secs)
+          }
+        };
+        format!("{}{}", duration_label, label)
+      })
+      .collect()
+  }
+
+  fn update_swim_lanes(
+    workunit_store: &WorkUnitStore,
+    tasks_to_display: &mut IndexMap<String, Option<Duration>>,
+    swim_lane_bars: &[ProgressBar],
+  ) {
+    let num_swimlanes = swim_lane_bars.len();
+    let heavy_hitters = workunit_store.heavy_hitters(num_swimlanes);
+
+    // Insert every one in the set of tasks to display.
+    // For tasks already here, the durations are overwritten.
+    tasks_to_display.extend(heavy_hitters.clone().into_iter());
+
+    // And remove the tasks that no longer should be there.
+    for (task, _) in tasks_to_display.clone().into_iter() {
+      if !heavy_hitters.contains_key(&task) {
+        tasks_to_display.swap_remove(&task);
+      }
+    }
+
+    let swimlane_labels: Vec<String> = Self::get_label_from_heavy_hitters(tasks_to_display.iter());
+    for (n, pbar) in swim_lane_bars.iter().enumerate() {
+      match swimlane_labels.get(n) {
+        Some(label) => pbar.set_message(label),
+        None => pbar.set_message(""),
+      }
+    }
+  }
+
+  pub fn render_loop(&self, refresh_interval: Duration) -> Result<Vec<RootResult>, String> {
+    loop {
+      let (display_sender, execution_status): (
+        mpsc::Sender<ConsoleMessage>,
+        mpsc::Receiver<ConsoleMessage>,
+      ) = mpsc::channel();
+      {
+        let mut mutex = self.display_sender.lock();
+        *mutex = Some(display_sender.clone());
+      }
+
+      let callback = move |msg: String| {
+        match display_sender.send(ConsoleMessage::Stderr(msg.clone())) {
+          Ok(()) => (),
+          Err(send_error) => {
+            eprintln!("Error writing message to console: {:?}. Original message: `{}`", send_error, msg);
+          }
+        }
+      };
+      LOGGER.set_console_display_callback(Box::new(callback));
+
+      let num_swimlanes = num_cpus::get();
+      let (multi_progress, bars) = Self::setup_bars(num_swimlanes);
+      let (display_loop_sender, display_loop_receiver) = mpsc::channel();
+
+      let workunit_store = self.workunit_store.clone();
+      let _ = std::thread::spawn(move || {
```

Comment:
Why do you spawn a thread here? I might be missing something, but it looks like you should be able to loop directly here, without the extra thread.

Reply:
You need the extra thread to interact with the `indicatif` library properly. Their example code (https://github.com/mitsuhiko/indicatif/blob/master/examples/multi.rs) assumes multithreading, and I couldn't figure out a way to update the screen reliably without spawning a new thread like this.